src/main/java/com/github/rschmitt/dynamicobject/internal/Instances.java

```diff
@@ -29,7 +29,20 @@
     private static <D extends DynamicObject<D>> D createIndyProxy(Map map, Class<D> type) {
         ensureInitialized(type);
         try {
-            Object proxy = proxyCache.computeIfAbsent(type, Instances::createProxy)
+            DynamicProxy dynamicProxy;
+            // Not using ConcurrentHashMap#computeIfAbsent to avoid locking: JDK-8161372
+            if (proxyCache.containsKey(type)) {
+                dynamicProxy = proxyCache.get(type);
+            } else {
+                synchronized (type) {
```

Comment:
I'm a little concerned about this synchronization actually - we don't own the object class, and don't know if some other thread might be entering this codepath with it already locked. In particular, static synchronized methods in user code will take these locks. Could this lead to lock inversion and deadlock?

I'd suggest going back to putIfAbsent or computeIfAbsent here, after the optimistic containsKey check.

Reply:
I can see a deadlock happening in the following scenario. One thread locks on class A and another thread on class B. Then the thread holding the lock on class A wants to deserialize a DynamicObject of type B. Likewise the second thread holding the lock on class B wants to deserialize an a DynamicObject of type A.

In this case, the putIfAbsent approach previously mentioned would likely be a middle ground, where we don't have class object synchronization and we don't use different proxies.

Reply:
I agree with @bdonlan, it's safer to go back to `putIfAbsent`.